salt-mode.el

```diff
@@ -194,6 +213,202 @@ module to use.
+When IF-MISSING is set, only refresh data that is empty."
+  (interactive)
+  (unless (and salt-mode--state-argspecs if-missing)
+    (salt-mode--with-async-minion "minion.functions.sys.state_argspec('*')"
+      (when (and result (hash-table-p result))
+        (setq salt-mode--state-argspecs result)
+        (message "Loaded %d Salt state function argument specifications."
+                 (hash-table-count result))))))
+
+(defun salt-mode--state-doc (module-or-function)
+  "Return documentation for the given state MODULE-OR-FUNCTION."
+  (let ((doc (gethash module-or-function
+                      (salt-mode--query-minion
+                       (format "minion.functions.sys.state_doc(%S)"
+                               module-or-function)))))
+    (when doc (replace-regexp-in-string "^    " "" doc))))
+
+(defun salt-mode--execution-doc (module-or-function)
+  "Return documentation for the given execution MODULE-OR-FUNCTION."
+  (gethash module-or-function
+           (salt-mode--query-minion
+            (format "minion.functions.sys.doc(%S)"
+                    module-or-function))))
+
+(defcustom salt-mode-hide-eldoc-argument-values
+  '(nil :json-false "")
+  "Default values to hide from the ElDoc function summary."
+  :group 'salt-mode
+  :type '(set (const :tag "Null" nil)
+              (const :tag "Empty string" "")
+              (const :tag "False" :json-false)
+              (const :tag "True" t)))
+
+(defconst salt-mode--false-string
+  (propertize "false" 'face 'font-lock-constant-face)
+  "String to use for showing false values.")
+
+(defconst salt-mode--true-string
+  (propertize "true" 'face 'font-lock-constant-face)
+  "String to use for showing true values.")
+
+(defconst salt-mode--null-string
+  (propertize "null" 'face 'font-lock-constant-face)
+  "String to use for showing null values.")
+
+(defconst salt-mode--mandatory-string
+  (propertize "?" 'face 'font-lock-keyword-face)
+  "String to use for showing true values.")
+
+(defconst salt-mode--kwargs-string
+  (propertize "... " 'face 'font-lock-keyword-face)
+  "String to use for unknown keyword arguments.")
+
+(defun salt-mode--format-argspec (name default)
+  "Format argument NAME with value DEFAULT for display."
+  (let ((name (propertize name 'face 'font-lock-variable-name-face)))
+    (if (member default salt-mode-hide-eldoc-argument-values)
+        name
+      (format "%s: %s" name
+              (cond ((eq default :json-false) salt-mode--false-string)
+                    ((eq default t) salt-mode--true-string)
+                    ((eq default :mandatory) salt-mode--mandatory-string)
+                    ((not default) salt-mode--null-string)
+                    (t (propertize
+                        (prin1-to-string default)
+                        'face 'font-lock-string-face)))))))
+
+(defun salt-mode--eldoc ()
+  "ElDoc support for salt-mode."
+  (when salt-mode--state-argspecs
+    (let* ((state-function (thing-at-point 'salt-mode-state-function))
```

Comment:
If you change this to `(salt-mode--state-module-at-point)`, eldoc will work anywhere within the state.

![screen shot 2017-06-20 at 22 43 04](https://user-images.githubusercontent.com/578458/27357437-0f46ba42-560a-11e7-8fc7-9cfac75c33cd.png)